### PR TITLE
[ds-fusion] Add RS->DUS dynamic slice fusion

### PR DESCRIPTION
**Pull request by shraiysh**
This patch adds execution support and fusion rewriting support for reduce-scatter plus slicing pattern. This includes slicing on operands or results of the reduce-scatter collective.

### CODE REVIEW

File: xla/tests/hlo_test_base.h [hunk 1]
```diff
@@ -432,6 +432,12 @@ class HloTestBase : public ManifestCheckingTest {
   static se::Platform* GetReferencePlatform();
   static se::Platform* GetTestPlatform();
 
+  // Compares the inputs shapes of two modules and returns the list of parameter
```
Comment (ezhulenev) [inline xla/tests/hlo_test_base.h:435]: Can you please split it into a separate change.

File: xla/service/gpu/dynamic_slice_fusion_rewriter.cc [hunk 2]
```diff
@@ -493,6 +496,10 @@ absl::StatusOr<bool> DynamicSliceFusionRewriter::Run(
       }
       TF_RETURN_IF_ERROR(
           parent->ReplaceInstruction(instr_to_be_replaced, fusion));
+      // This is required for collective operations which will not be removed.
+      if (hero->parent() && hero->user_count() == 0) {
+        TF_RETURN_IF_ERROR(hero->parent()->RemoveInstruction(hero));
```
Comment (ezhulenev) [inline xla/service/gpu/dynamic_slice_fusion_rewriter.cc:501]: I think it is an error if after the rewrite `user_count` is not 0, and it should be safe to always remove instruction explicitly.

**Review by ezhulenev — APPROVED** [review]

**Comment by mooskagh** [comment]
The added tests require 8 GPUs to run, but Google testing fleet has 2 GPU machines at most (that's why tests fail).

Would it make sense/be easy for you to change the test to use 2 GPUs only?

**Comment by shraiysh** [comment]
>
> The added tests require 8 GPUs to run, but Google testing fleet has 2 GPU machines at most (that's why tests fail).
>
> Would it make sense/be easy for you to change the test to use 2 GPUs only?

Done!

File: xla/tests/hlo_test_base.h (hunk 1 above)
Comment (akuegel) [reply xla/tests/hlo_test_base.h:435]: @shraiysh it seems this comment is not addressed yet? It is usually better to do such refactorings separately as they are no functional change, so less likely that they need to be rolled back, and it makes PRs smaller and easier to review.

Comment (shraiysh) [reply xla/tests/hlo_test_base.h:435]: Added #14822

File: xla/service/gpu/fusions/custom.cc [hunk 3]
```diff
@@ -812,10 +813,136 @@ absl::StatusOr<FusionEmissionResult> CustomFusion::Emit(
...
+      .destination_value = nullptr});
+
+  if (instr->opcode() == HloOpcode::kReduceScatter) {
+    int64_t replica_count = instr->GetModule()->config().replica_count();
+    int64_t partition_count = instr->GetModule()->config().num_partitions();
+    auto rs = static_cast<const HloReduceScatterInstruction*>(instr);
+    TF_RETURN_IF_ERROR(NcclReduceScatterStartThunk::CheckImplementable(
+        rs, replica_count, partition_count));
+
+    // TODO: add special handling for degenrate case - where no communication is
```
Comment (akuegel) [inline xla/service/gpu/fusions/custom.cc:908]: nit: degenrate -> degenerate

**Review by akuegel — COMMENTED** [review]
Thanks for landing the test refactoring separately. Just found a minor typo

**Review by akuegel — APPROVED** [review]

**Comment by shraiysh** [comment]
@akuegel can you mention what the internal failure says?

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
> @akuegel can you mention what the internal failure says?

I also don't see it. I think it happened before the import has finished, so I don't see a corresponding internal change. I approved again, maybe it will work this time? Another thing worth trying is if you can rebase this PR, maybe that helps as well.

File: xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc [hunk 4]
```diff
@@ -1752,4 +1752,136 @@ TEST_F(DynamicSliceFusionRewriterTest, DUSSimpleGemmWorkspaceIgnored) {
...
+  // CHECK: }
+  )";
+  RunAndFilecheckHloRewrite(hlo, DynamicSliceFusionRewriter("gpu"),
+                            expected);
+}
+
+// This is not required to pass, but in current implementation this works by
+// forcing a D2H copy. Adding this here to ensure that the change in this
+// behaviour is intentional.
+TEST_F(DynamicSliceFusionRewriterTest, ReduceScatterDUSParameterOffset) {
```
Comment (frgossen) [inline xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc:1792]: Can you simplify the test cases? 
They look like JAX dumps, e.g. metadata is not relevant for these tests.

**Review by frgossen — CHANGES_REQUESTED** [review]
> This patch adds execution support and fusion rewriting support for reduce-scatter -> dynamic-update-slice pattern.

Can you describe in a bit more detail what this PR is doing? 
What do you mean with "reduce-scatter -> dynamic-update-slice pattern"? 

Can you explain the benefit of this CL? Do you observe performance improvements on a specific benchmark?

**Comment by thomasjoerg** [comment]
@shraiysh while integrating this change I noticed a broken test. `DynamicSliceFusionTest.ReduceScatterDUSLoopIterationOffset` fails with 

```
Expected equality of these values:
  address_computations_ref.size()
    Which is: 1
  0
```

Can you reproduce this issue? You may need to sync to head.

**Comment by thomasjoerg** [comment]
@shraiysh friendly ping!

File: xla/service/gpu/fusions/custom.cc [hunk 5]
```diff
@@ -812,10 +813,136 @@ absl::StatusOr<FusionEmissionResult> CustomFusion::Emit(
...
+  BufferAllocation::Slice slice_out_fake(fake_allocations[fake_arg_idx].get(),
+                                         0, out_fake_byte_size);
+
+  // Generate the hero thunk and wrap it in a dynamic-slice thunk.
+  ThunkSequence seq;
+  auto thunk_info = Thunk::ThunkInfo::WithProfileAnnotation(instr);
+  std::vector<NcclCollectiveThunk::Buffer> buffers;
+  const Shape& src_shape = instr->operand(0)->shape();
+  const Shape& dst_shape = instr->shape();
+  buffers.push_back(NcclCollectiveThunk::Buffer{
```
Comment (thomasjoerg) [inline xla/service/gpu/fusions/custom.cc:892]: Please don't use C++20: Designated initializers are incompatible with C++ standards before C++20

**Review by thomasjoerg — CHANGES_REQUESTED** [review]

File: xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc [hunk 6]
```diff
@@ -1752,4 +1752,136 @@ TEST_F(DynamicSliceFusionRewriterTest, DUSSimpleGemmWorkspaceIgnored) {
...
+  // CHECK: ENTRY {{.+}} {
+  // CHECK-NOT: reduce-scatter
+  // CHECK:   ROOT %{{.+}} = {{.+}} fusion(%{{.+}}, %{{.+}}, %{{.+}}, %{{.+}}), kind=kCustom, calls=%address-computation, {{.+}}"fusion_backend_config":{"kind":"__custom_fusion","custom_fusion_config":{"name":"dynamic_address_computation"}}
+  // CHECK: }
+  )";
+  RunAndFilecheckHloRewrite(hlo, DynamicSliceFusionRewriter("gpu"),
+                            expected);
+}
+
+TEST_F(DynamicSliceFusionRewriterTest, ReduceScatterDUSLoopIterationOffset) {
```
Comment (frgossen) [inline xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc:1825]: This test case still seems to be unnecessarily verbose.

**Comment by shraiysh** [comment]
> > This patch adds execution support and fusion rewriting support for reduce-scatter -> dynamic-update-slice pattern.
> 
> Can you describe in a bit more detail what this PR is doing? What do you mean with "reduce-scatter -> dynamic-update-slice pattern"?
> 
> Can you explain the benefit of this CL? Do you observe performance improvements on a specific benchmark?

I believe I should rephrase the pattern to be `reduce-scatter` plus any slicing operations on operands or results. This handles slicing patterns on both operands and results.

I do not have number for a specific benchmark for this, but this change stemmed from the [document](https://docs.google.com/document/d/13gVd5CFe_meK94iOip1CB5pAodyN4ce2pL0hfucEgLg/edit#heading=h.x9snb54sjlu9) that @ezhulenev wrote about address computation fusion. Let me know if you would like concrete benchmark numbers with this.

File: xla/service/gpu/fusions/custom.cc (hunk 5 above)
Comment (shraiysh) [reply xla/service/gpu/fusions/custom.cc:835]: Done. Thanks for pointing this out.

File: xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc (hunk 6 above)
Comment (shraiysh) [reply xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc:1823]: I have removed metadata instructions. Do you think I should remove some other stuff too?

File: xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc (hunk 4 above)
Comment (shraiysh) [reply xla/service/gpu/dynamic_slice_fusion_rewriter_test.cc:1791]: Done. Thanks!

**Comment by shraiysh** [comment]
>@shraiysh while integrating this change I noticed a broken test. DynamicSliceFusionTest.ReduceScatterDUSLoopIterationOffset fails

I believe I have handled the issue with the rebase. Can you check now? Apologies for the delay in this work.

**Review by thomasjoerg — APPROVED** [review]

**Review by Flako125 — APPROVED** [review]

**Pull request closed without merging**